### DISCUSSION TITLE
Algotrithms used in automatic tree layout in JointJS+

### DISCUSSION THREAD

**Discussion by hpietukhin** [clientIO/joint#2375] [Q&A] [answered]
### Introduction

I want to use JointJS library in my undergraduate thesis. I want to know what algorithms are used in JointJS+ for automatic tree layout?
I suppose that [TreeLayout](https://resources.jointjs.com/docs/rappid/v3.7/layout.TreeLayout.html#layout.TreeLayout) is just extended [DirectedGraphLayout](https://resources.jointjs.com/docs/jointjs/v3.7/joint.html#layout.DirectedGraph), which internally uses works described in section [Recommended reading](https://github.com/dagrejs/dagre/wiki#recommended-reading).

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS+

**Comment by kumilingus** [comment, marked as answer]
The `TreeLayout` is not an extension of `DirectedGraphLayout`. It is a different algorithm. First you go from the leaves to the root and find relative positions and then you go from the root to the leaves and set the absolute positions.

Here's a nice article on tree visualization: https://www.zxch3n.com/tidy/tidy/
